### PR TITLE
Clang headers aren't extracted in a static folder anymore

### LINKED ISSUE #517: Create a temporary unique folder for clang-headers extraction
Using a static name is causing data-race and permission issues when running esbmc with multiple users

### PR DESCRIPTION
**Pull request by rafaelsamenezes**
This PR:

1. Creates the file_operations class which should contain helper functions for file/folder modules.
2. Added a method that creates an unique folder inside the temp dir, using a format string as a base
3. Refactored clang-c-language and message to use the new lib

Closes #517

### CODE REVIEW

**Review by lucasccordeiro — APPROVED** [review]
Could you create a test case?

File: src/clang-c-frontend/clang_c_language.cpp [hunk 1]
```diff
@@ -32,29 +33,16 @@ languaget *new_clang_c_language(const messaget &msg)
...
-  if(!boost::filesystem::is_directory(p))
-  {
-    msg.error(
-      "Can't create temporary directory (needed to dump clang headers)");
-    abort();
-  }
-
+  /* About the path being static:
+   * the function dump_clang_headers has a static member checking if it was
+   * ever extracted before. This will garantee that the same path will be used
```
Comment (lucasccordeiro) [inline src/clang-c-frontend/clang_c_language.cpp]: "garantee" -> "guarantee"

File: src/clang-c-frontend/clang_c_language.cpp [hunk 2]
```diff
@@ -32,29 +33,16 @@ languaget *new_clang_c_language(const messaget &msg)
...
-      "Can't create temporary directory (needed to dump clang headers)");
-    abort();
-  }
-
+  /* About the path being static:
+   * the function dump_clang_headers has a static member checking if it was
+   * ever extracted before. This will guarantee that the same path will be used
+   * during a run. And no more than one is required anyway */
+  static auto p =
+    file_operations::get_unique_tmp_path("esbmc-headers-%%%%-%%%%-%%%%");
```
Comment (mikhailramalho) [inline src/clang-c-frontend/clang_c_language.cpp:41]: nice

File: src/util/filesystem.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+   *
+   * In Linux, running this function with "esbmc-%%%%" will
+   * return a string such as "/tmp/esbmc-0001" or "/tmp/esbmc-8787".
+   *
+   * This function does not have guarantee that will finish
+   * and can be run forever until it sees an available spot.
+   *
+   * @param format A string in the file specification
+   */
+  static std::string get_unique_tmp_path(const char *format);
```
Comment (mikhailramalho) [inline src/util/filesystem.h]: const std::string?

**Review by mikhailramalho — APPROVED** [review]
LGTM.

one nit only: do we really need a `const char*` there? Can't we use string?

**Pull request merged**